controller/query.go

```diff
@@ -0,0 +1,147 @@
+package controller
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/fabric8-services/fabric8-wit/app"
+	"github.com/fabric8-services/fabric8-wit/application"
+	"github.com/fabric8-services/fabric8-wit/jsonapi"
+	"github.com/fabric8-services/fabric8-wit/login"
+	"github.com/fabric8-services/fabric8-wit/query"
+	"github.com/fabric8-services/fabric8-wit/rest"
+	"github.com/fabric8-services/fabric8-wit/search"
+	"github.com/goadesign/goa"
+	"github.com/prometheus/common/log"
+)
+
+// QueryController implements the query resource.
+type QueryController struct {
+	*goa.Controller
+	db     application.DB
+	config QueryControllerConfiguration
+}
+
+// QueryControllerConfiguration the configuration for the LabelController
+type QueryControllerConfiguration interface {
+	GetCacheControlQueries() string
+	GetCacheControlQuery() string
+}
+
+// NewQueryController creates a query controller.
+func NewQueryController(service *goa.Service, db application.DB, config QueryControllerConfiguration) *QueryController {
+	return &QueryController{
+		Controller: service.NewController("QueryController"),
+		db:         db,
+		config:     config,
+	}
+}
+
+// Create runs the create action.
+func (c *QueryController) Create(ctx *app.CreateQueryContext) error {
+	currentUserIdentityID, err := login.ContextIdentity(ctx)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+	return application.Transactional(c.db, func(appl application.Application) error {
+		q := query.Query{
+			SpaceID: ctx.SpaceID,
+			Fields:  ctx.Payload.Data.Attributes.Fields,
+			Title:   strings.TrimSpace(ctx.Payload.Data.Attributes.Title),
+			Creator: *currentUserIdentityID,
+		}
+		// Parse fields to make sure that query is valid
+		_, _, err := search.ParseFilterString(ctx, q.Fields)
```

Comment:
The `err` value only indicate whether the query is syntactically correct.  Where as the first return value is of type `criteria.Expression` and if that value is `nil`, the expression is still invalid. Should we catch that?

Reply:
thanks @baijum I wasn't aware of it, done in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/2c7f2d8b82dfd68ccfa159a7e6373bfe991266c4